Refactor Coordinape Integration

mckelvey opened the pull request:
This PR updates the Coordinape integration into the ETL format. Major items:

1. It loads past data from the community, and updates it, failing in the event the data has changed in the interim.
2. It throttles Piñata and IPFS gateways requests.
3. It updates the community dashboard to force connecting and creating prior to integration.
4. It updates the API authentication with a custom error.

If this looks good, it does force some changes for the other integrations and might allow for some shared actions, such as how past community data is loaded. I’ll also need to remove some of the older files.

## Questions Worth a Discussion

1. When we are unable to update and store member data to IPFS, what should we do? Kill the whole cycle?
2. Is it the right approach to recheck and kill the community transaction if the community hash has changed, or just restart the whole thing?
3. Could this be less prone to update conflicts by tracking member data attached to their Fauna record and not in the community? Would that have any impacts on rollups?

## CSV Notes

There were some difficulties in handling the data. Initially, I’d intended to upload the files in bulk via `FormData` but then found that I was unable to get nextjs to wait for parsing the incoming data. And this was with a custom webpack build in the next config just to be able to handle the lack of `fs` which any of `formidable` or `multiparty` both seemed to require.

In the end, I gave up and instead accepted that it might be a more reasonable approach to send the data as JSON, since any future Coordinape API would almost certainly be an epoch-aligned API request, and this would keep the ETL closer to what we might need longer-term.

But that was not the end, as any CSV parser either wanted either `fs` or streams and the reality is that these Coordinape CSV files are generally small and consistently formed. (Think about it, their UI wouldn’t really support a circle of hundreds nicely, so we can count on smaller circles and reasonable files that we can read into memory.) Soooo, I wrote my own Coordinape CSV parser, which also handles casting integers and naming. It doesn’t require either `fs` or `streams` and so no nextjs mods as needed for tools we won’t be using anyway.

royriver23 commented on .env.example:
```diff
@@ -1,8 +1,21 @@
+RINKEBY_RPC_PATH=alchemypath
+MAINNET_RPC_PATH=alchemypath
+PINATA_PUBLIC_API_KEY=8c43bf70fe8da866dc10
+PINATA_PRIVATE_API_KEY=58d493e409266266cfce10710ad16dd27ccd7f4dc12b02d0efe20c82650b58b3
```
expected to have this in here?

royriver23 commented on components/integrations/view/coordinape/IntegrationForm.js:
```diff
@@ -0,0 +1,93 @@
+          body: JSON.stringify({
+            name: file.name,
+            contributions: coordinapeData,
+          }),
```
overkill to move into hook?

mckelvey replied:
Probably not in this case, as we shouldn’t have a lot of these where we are uploading CSVs (I hope).

royriver23 commented on pages/community-dashboard.js:
```diff
@@ -1,19 +1,53 @@
+  useEffect(() => {
+    // if connected, try to load community
+    if (address && step === STEPS.CONNECT) {
+      setStep(STEPS.LOAD)
```
oooh, these step-based effects are awesome!

royriver23 commented on util/api/withAuth.js:
```diff
@@ -1,11 +1,21 @@
 import jwt from 'jsonwebtoken'
 import { utils } from 'ethers'
 
+class AuthenticationError extends Error {
```
Niceeee

royriver23 commented on util/csv.js:
```diff
@@ -0,0 +1,85 @@
+
+const knownHeaderKeys = Object.keys(headerKeyMap)
+
+const asPassThru = value => value
```
self-documented, niceee

royriver23 commented on util/integrations/coordinape/etl.js:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  // Extract data:
+  // const { name } = epochData
```
Same to remove, or leaving it for documentation reasons?

mckelvey replied:
I have a little cleanup here and elsewhere yet I think.

royriver23 commented on util/integrations/coordinape/load.js:
```diff
@@ -0,0 +1,56 @@
+      upsertMemberData(data[address], members[address]),
+    ),
+  )
+
```
We could add some error handling to at least detect failing ones and log them?

```javascript
const updatedIpfsHashes = await Promise.allSettled(
    incomingDataAddresses.map(async address =>
      upsertMemberData(data[address], members[address]),
    ),
  )
```

Then, we can map results and see the ones failing? by comparing against `incomingDataAddresses` array?

royriver23 commented on util/ipfs.js:
```diff
@@ -0,0 +1,20 @@
+
+const rateLimit = parseInt(process.env.IPFS_GATEWAY_RATE_LIMIT, 10)
+
+const gateway = new Bottleneck({
```
WOOOW!

nicovalencia replied:
Wonder if this applies to a paid/pro gateway. I think we go around that on other projects using a paid gateway.

royriver23 reviewed (approved):
Nice work, David! Some comments, but they aren't crucial!

mckelvey commented on util/queries.js:
```diff
@@ -39,9 +39,13 @@ export const FIND_COMMUNITY_BY_ADDRESS_QUERY = gql`
 export const UPDATE_COMMUNITY_DATA_HASH_MUTATION = gql`
   mutation UPDATE_COMMUNITY_DATA_HASH_MUTATION(
     $id: ID!
+    $address: String!
```
@nicovalencia — I had to add this to get Fauna to save it. If `address` is unique and indexed, should we be using the `_id`, or conversely, this means the same address could never admin more than one community. (But perhaps that’s OK as we expect this to be a gnosis safe or other shared address?)

mckelvey replied:
Totally agree, but when I removed ID, it wouldn’t update... This might well be my lack of experience with Fauna, so if there’s some guidance on this, I’ll take it. :) I also figured it could be a followup PR to remove it.

royriver23 reviewed (commented):
> 1. When we are unable to update and store member data to IPFS, what should we do? Kill the whole cycle?

Good question! I think we should proceed with the ones we could, and collect the addresses for the ones we couldn't, then make a decision based on the following:

* When retryable error: we could retry or ask community owner to retry (re-upload file or use cached uploaded one)
* When non-retryable error: we should inform the list of addresses along with the failure reason.

> 2. Is it the right approach to recheck and kill the community transaction if the community hash has changed, or just restart the whole thing?

Not sure, if following, are you referring to `throw new Error('Another ETL was running simultaneously; please try again.')`?  If so, I think we could restart the whole thing after having the community owner confirm the overwrite. Just killing works too, since the error message you chose is meaningful.

> 3. Could this be less prone to update conflicts by tracking member data attached to their Fauna record and not in the community? Would that have any impacts on rollups?

Wow, that makes sense to me, we could manage hash at member level, then use the array of member's hashes as differentiator. Maybe just having an extra field on community level, and the current hashing logic does the rest, as it was yet another community field to take into account. As for the rollups, since this is the data loading step, I'm not sure how it would affect.

@mckelvey , of course, happy to see others' opinions, cc @nicovalencia

mckelvey commented:
This PR is now ready for review. These are the changes since the previous draft:

1. Error handling has been removed so that any failure fails the entire ETL. (@royriver23, @nicovalencia)
2. Community and member Fauna and IPFS data handling is DRY(ed) up into data helpers that facilitate a unified, non-destructive approach to DB/IPFS storage. (@nicovalencia)
3. Snapshot ETL and example ETL are now updated to use the community/member data helpers.
4. The previous Coordinape approach has been removed.

There is an issue with GitHub actions installing the needed libs that is causing failures of the checks when run by GitHub, but local runs of linting, tests and Chromatic are fine.

nicovalencia commented on util/member.js:
```diff
@@ -0,0 +1,63 @@
+    }),
+    membersDataHashes,
+  )
+}
```
Thinking it might be nice to nest API specific util files under `utill/api/*` thoughts?

The other day someone accidentally imported a util meant for the API in the client side and it broke because it was expecting private runtime values.

mckelvey replied:
Good point! Will move.

mckelvey commented:
Additional note on the GitHub actions fails— it looks like this is an issue with a corrupted `package.lock`; I’ll be looking at means to fix that without `rm` if possible, since that might cause more issues than it solves.

mckelvey commented on util/integrations/coordinape/transform.js:
```diff
@@ -0,0 +1,48 @@
+    const { address: downcaseAddress, received, sent } = contribution
+
+    // Coordinape CSV addresses are downcased, we prefer Checksum
+    const address = utils.toChecksumAddress(downcaseAddress)
```
@digidigo — Would you believe this? Somehow, I’d have thought Coordinape would be a Checksum crew. I guess not.

mckelvey commented:
NOTE: I had to skip a contract test as it is failing in develop... I’ll follow up with a task to address it if/when this is merged.

royriver23 commented on .env.example:
```diff
@@ -1,8 +1,21 @@
+NEXT_PUBLIC_TWITTER_USERNAME='lvlprotocol'
+NEXT_PUBLIC_ALCHEMY_KEY=key
+NEXT_PUBLIC_FAUNA_URL=https://graphql.us.fauna.com/graphql
+NEXT_PUBLIC_GITHUB_OAUTH_CLIENT_ID=client_id
```
do we really need these if we are already falling back to `.env.development` ?

mckelvey replied:
@royriver23 and I chatted about this and I’ve updated the `.env.example` with our resolved solution. :)

royriver23 reviewed (approved):
niceeee

royriver23 reviewed (approved):
<img src="https://media4.giphy.com/media/3orif469YgEH4oeVIA/giphy.gif"/>
